### PR TITLE
plan: adjust index cost calculation algorithm.

### PR DESCRIPTION
**Pull request by hanfei1991**
1. When the conditions are "a in (1,2) and b in (3,4)", the count will be rounded to 0, which will cause terrible bug.
2. When the selection is pushed down and the prop has limit, we should recalculate cost for table scan and index scan.
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu PTAL

### CODE REVIEW

File: plan/statistics/statistics.go [hunk 1]
```diff
@@ -31,10 +31,10 @@ const (
 	defaultBucketCount = 256
 
 	// When we haven't analyzed a table, we use pseudo statistics to estimate costs.
-	// It has row count 10000, equal condition selects 1/10 of total rows, less condition selects 1/3 of total rows,
+	// It has row count 1000000000, equal condition selects 1/500 of total rows, less condition selects 1/3 of total rows,
 	// between condition selects 1/4 of total rows.
-	pseudoRowCount    = 10000
-	pseudoEqualRate   = 1000
+	pseudoRowCount    = 1000000000
```
Comment (coocood) [inline plan/statistics/statistics.go]: It's an unreasonable number, makes code much harder to read.

File: plan/statistics/statistics.go [hunk 2]
```diff
@@ -31,9 +31,9 @@ const (
 	defaultBucketCount = 256
 
 	// When we haven't analyzed a table, we use pseudo statistics to estimate costs.
-	// It has row count 10000, equal condition selects 1/10 of total rows, less condition selects 1/3 of total rows,
+	// It has row count 1000000000, equal condition selects 1/1000 of total rows, less condition selects 1/3 of total rows,
 	// between condition selects 1/4 of total rows.
-	pseudoRowCount    = 10000
+	pseudoRowCount    = 1000000000
```
Comment (coocood) [inline plan/statistics/statistics.go]: It's an unreasonable number, makes code much harder to read.

File: plan/physical_plan_builder.go [hunk 3]
```diff
@@ -373,6 +384,10 @@ func enforceProperty(prop *requiredProperty, info *physicalPlanInfo) *physicalPl
 }
 
 func sortCost(cnt uint64) float64 {
+	if cnt == 0 {
+		// If cnt is 0, the log(cnt) will be NAN.
+		cnt = 10
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: Why it is `10` ？

File: plan/physical_plan_builder.go [hunk 4]
```diff
@@ -75,8 +77,21 @@ func getRowCountByIndexRange(table *statistics.Table, indexRange *IndexRange, in
...
+		// so as to avoid collapsing too fast.
+		for j := 0; j < i; j++ {
+			count = count / float64(100)
+		}
+		totalCount += count
+	}
+	// To avoid the totalCount become 0.
+	if uint64(totalCount) == 0 {
+		totalCount = 10.0
+	}
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: This code is also used to line 388, it can extract into a function. And use the same comments it will be clearer.

**Comment by coocood** [comment]
Please only include changes necessary to fix `Log2(0)` produce negative cost but.
We can discuss the cost model change later in future PR.

File: plan/physical_plan_builder.go (hunk 3 above)
Comment (coocood) [inline plan/physical_plan_builder.go]: I think this is wrong algorithm.
If cnt == 0, we should just return 0.

File: plan/physical_plan_builder.go [hunk 5]
```diff
@@ -85,8 +85,9 @@ func getRowCountByIndexRanges(table *statistics.Table, indexRanges []*IndexRange
 		totalCount += count
 	}
 	// To avoid the totalCount become 0.
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: Update comments

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**